Multi-domain logins for Windows Desktop

probakowski opened the pull request:
This change allows users to login to Windows desktop using multiple different domains as long as the trust is setup between them and LDAP returns proper referrals for the other domain.
If you have `domain1` configured in Teleport and `domain2` that `domain1` trusts then loggins as `user1` will be treated as `user1@domain1`, and `user1@domain1` and `user2@domain2` will use users from respective domains (and still log you in to the requested desktop in the `domain1`).

changelog: Add multi-domain logins for Windows Desktop

probakowski committed:
Multi-domain logins for Windows Desktop

zmb3 commented on lib/auth/windows/ldap.go:
```diff
@@ -218,7 +273,21 @@ func (c *LDAPClient) ReadWithFilter(dn string, filter string, attrs []string) ([
+				}
+			}
+		}
+		return nil, trace.Wrap(fmt.Errorf("no referral provided by LDAP server can execute the query, tried: %s", strings.Join(referrals, ",")))
```
I would just use trace.BadParameter here

zmb3 commented on lib/auth/windows/ldap.go:
```diff
@@ -167,6 +181,47 @@ func (c *LDAPClient) Close() {
+	out := make([]string, 0, len(referrals))
+	for _, referral := range referrals {
+		referralValue := referral.Value.(string)
+		last := strings.LastIndex(referralValue, "/")
```
This might be a little cleaner with strings.Cut. Give it a try and see if you like it better.

probakowski replied:
`string.Cut` cuts after first separator, we have to cut after last

zmb3 commented on lib/auth/windows/ldap.go:
```diff
@@ -22,6 +22,8 @@ import (
 	"crypto/x509"
 	"errors"
 	"fmt"
+	"github.com/go-asn1-ber/asn1-ber"
```
Do we need this?

probakowski replied:
Yes, it's used in `extractReferrals` (`packet *ber.Packet`)

zmb3 replied:
Does Go's `encoding/asn1` package work instead?

kopiczko replied:
For my curiosity: is it answered somewhere?

zmb3 commented on lib/auth/windows/ldap.go:
```diff
@@ -167,6 +181,47 @@ func (c *LDAPClient) Close() {
 	c.mu.Unlock()
 }
 
+func extractReferrals(err error) []string {
```
I'd add some godocs to this one. 

Maybe even make it accept an LDAPError instead of a generic error since it can't operate on any other type of error.

probakowski replied:
I've added some documentation and changed arg to `LDAPError`

zmb3 commented on lib/srv/desktop/windows_server.go:
```diff
@@ -591,6 +594,7 @@ func (s *WindowsService) initializeLDAP() error {
 	}
 
 	conn.SetTimeout(ldapRequestTimeout)
+	s.lc.SetConnectionCreator(createLDAPConnection)
```
Why do we need this connection creator thing? (Instead of just using ldap.Dial directly)

probakowski replied:
We set up TLS here but we create new connection inside `ldap.ReadWithFilter` (when we get new referral to check in the response). This way we have factory for these connection with certificates and so on

kopiczko replied:
AFAICT it's called only in `WindowsService.initializeLDAP`, so is `SetClient`. Feels like `s.lc` could be created here instead and then we could avoid those setters and mutex. `LDAPConfig` is available in `s.cfg.LDAPConfig` anyway. Chances are I'm missing something here.

probakowski replied:
We pass `s.lc` to in `NewWindowsService` (`lib/srv/desktop/windows_server.go:387`) before `initializeLDAP`. We could probably rework that but it'd be awkward. I'd rather leave it as is for now.

probakowski committed:
review comments

probakowski committed:
lint

kopiczko commented on go.mod:
```diff
@@ -244,6 +244,8 @@ require (
 	software.sslmate.com/src/go-pkcs12 v0.5.0
 )
 
+require github.com/go-asn1-ber/asn1-ber v1.5.7
```
Perhaps should be included in the require block above.

kopiczko commented on lib/auth/windows/ldap.go:
```diff
@@ -218,7 +275,30 @@ func (c *LDAPClient) ReadWithFilter(dn string, filter string, attrs []string) ([
+		return res.Entries, nil
+	}
+
+	var ldapErr *ldap.Error
+	if !errors.As(err, &ldapErr) {
+		return nil, trace.Wrap(err, "fetching LDAP object %q with filter %q", dn, filter)
+	}
```
Maybe it would be cleaner to do this check inside `extractReferrals`, so we'd avoid writing the error string twice.

probakowski replied:
I refactored it a bit to avoid writing the same thing twice

kopiczko commented on lib/auth/windows/windows.go:
```diff
@@ -327,7 +333,7 @@ var EnhancedKeyUsageExtension = pkix.Extension{
 
 // SubjectAltNameExtension fills in the SAN for a Windows certificate
-func SubjectAltNameExtension(user, domain string) (pkix.Extension, error) {
+func SubjectAltNameExtension(userPrincipalName string) (pkix.Extension, error) {
```
Any reason this isn't private?

probakowski replied:
Nope, it's private now

probakowski committed:
gci

probakowski committed:
Merge remote-tracking branch 'origin/master' into probakowski/multi-domain

probakowski committed:
go.mod

probakowski committed:
review comments

probakowski committed:
fix test

probakowski committed:
fix test

probakowski committed:
go.mod

probakowski committed:
gci

probakowski committed:
gci

probakowski committed:
Merge remote-tracking branch 'origin/master' into probakowski/multi-domain

zmb3 committed:
discovery: try to determine the domain from the LDAP entry

We now expect that users may be discovering hosts in a different
(child) domain than the one Teleport is pointed at, so compute the
domain from the LDAP attributes rather than assume it to be the
same as the one Teleport is configured for.

probakowski committed:
Update e

probakowski committed:
Merge remote-tracking branch 'origin/master' into probakowski/multi-domain

probakowski committed:
Add back referrals

probakowski commented on go.mod:
```diff
@@ -269,6 +269,8 @@ require (
 	software.sslmate.com/src/go-pkcs12 v0.6.0
 )
 
+require github.com/go-asn1-ber/asn1-ber v1.5.8-0.20250403174932-29230038a667
```
Yep, I moved it.

probakowski commented on lib/winpki/ldap.go:
```diff
@@ -215,7 +269,41 @@ func (l *LDAPClient) ReadWithFilter(dn string, filter string, attrs []string) ([
+	}
+
+	var ldapErr *ldap.Error
+	if errors.As(err, &ldapErr) && ldapErr.ResultCode == ldap.LDAPResultReferral {
```
I refactored the code to minimize nesting.

probakowski commented on lib/winpki/ldap.go:
```diff
@@ -215,7 +269,41 @@ func (l *LDAPClient) ReadWithFilter(dn string, filter string, attrs []string) ([
+		for i := 0; i < len(referrals); i++ {
+			cfg := LDAPConfig{
+				Addr:     referrals[i],
+				Domain:   l.cfg.Domain,
```
No, in fact we don't need to set it at all (it's used only when locating server through DNS)

probakowski commented on lib/winpki/ldap.go:
```diff
@@ -215,7 +269,41 @@ func (l *LDAPClient) ReadWithFilter(dn string, filter string, attrs []string) ([
+				res, err := conn.SearchWithPaging(req, searchPageSize)
+				if err == nil {
+					return res.Entries, nil
+				} else if len(referrals) < 10 && errors.As(err, &ldapErr) && ldapErr.ResultCode == ldap.LDAPResultReferral {
```
I put wrong address in AD and created referrals loop once - this is safeguard against that.

probakowski commented on lib/winpki/ldap.go:
```diff
@@ -199,9 +209,53 @@ func (l *LDAPClient) GetActiveDirectorySID(ctx context.Context, username string)
+
+	out := make([]string, 0, len(referrals))
+	for _, referral := range referrals {
+		referralValue := referral.Value.(string)
```
Added `, ok` check.

probakowski commented on lib/winpki/ldap.go:
```diff
@@ -199,9 +209,53 @@ func (l *LDAPClient) GetActiveDirectorySID(ctx context.Context, username string)
 }
 
+// extractReferrals gathers referrals from ldapErr
+// If LDAP server can't provide the information required but has the knowledge of proper it will return error like:
```
"of proper server" - I reworded this comment

probakowski committed:
Merge remote-tracking branch 'origin/master' into probakowski/multi-domain

probakowski commented on lib/srv/desktop/discovery.go:
```diff
@@ -212,6 +213,8 @@ func (s *WindowsService) applyLabelsFromLDAP(entry *ldap.Entry, labels map[strin
 		labels[types.DiscoveryLabelWindowsOU] = ou
 	}
 
+	labels[types.DiscoveryLabelWindowsDomain] = dnToDomain(dn)
```
We do set it but with multiple domains you can configure discovery for whole forest so domain should reflect the real one

probakowski commented on lib/winpki/ldap.go:
```diff
@@ -199,9 +209,53 @@ func (l *LDAPClient) GetActiveDirectorySID(ctx context.Context, username string)
+		return nil
+	}
+	searchResultIndex := slices.IndexFunc(ldapErr.Packet.Children, func(packet *ber.Packet) bool {
+		return packet.Description == "Search Result Done"
```
Unfortunately no, there's no id of any kind, just description. `go-ldap` hardcodes indices so it looks like packets always have the same structure but I haven't found confirmation of this anywhere in the docs, this way is more robust.
I did so refactoring of the code above, btw.

probakowski committed:
SID LDAP query to use both sAMAccountName and UPN

probakowski committed:
Merge remote-tracking branch 'origin/master' into probakowski/multi-domain

probakowski committed:
Merge remote-tracking branch 'origin/master' into probakowski/multi-domain

probakowski committed:
New queries and DN

danielashare committed:
Don't check for KDC address in non-AD environments

probakowski committed:
Merge remote-tracking branch 'origin/master' into probakowski/multi-domain

probakowski committed:
fix test

probakowski committed:
Merge remote-tracking branch 'origin/master' into probakowski/multi-domain

probakowski committed:
cleanup
